File: SEMESTER3/PemrogramanBerorientasiObjek/Week10/TUGAS/Inheritance.cpp
/*
* Nama          : Amir Salim , Andre Nathaniel Adipraja , Prames Ray lapian , Ibrahim Dafi Iskandar
* NPM           : 140810210015 , 140810200042 , 140810210059 , 140810210039
* Kelas         : A
* Tanggal       : 11 November 2022
* Nama Program  : Inheritance.cpp
* Deskripsi     : UTS SOAL 4 dengan Inheritance
*/

#include<iostream>
#include<string>


class Waktu{
    private:
        int jam,menit,detik;
    
    public:
        Waktu(int jam,int menit,int detik){
            this->jam = jam;
            this->menit = menit;
            this->detik = detik;
        }
        Waktu(){
            this->jam = 0;
            this->menit =0;
            this->detik=0;
        }
        void setJam(int jam){
            this->jam = jam;
        }
        void setMenit(int menit){
            this->menit = menit;
        }
        void setDetik(int detik){
            this->detik = detik;
        }
        void inputWaktu(){
            std::cout<<"Masukkan jam : ";
            std::cin>>this->jam;

            std::cout<<"Masukkan menit : ";
            std::cin>>this->menit;

            std::cout<<"Masukkan detik : ";
            std::cin>>this->detik;
        }
        int getJam(){
            return this->jam;
        }
        int getMenit(){
            return this->menit;
        }
        int getDetik(){
            return this->detik;
        }
        std::string getWaktu(){
            std::string nolJam ="";
            std::string nolMenit="";
            std::string nolDetik="";

            if(this->jam<10){
                nolJam="0";
            }
            if(this->menit<10){
                nolMenit="0";
            }
            if(this->detik<10){
                nolDetik="0";
            }
            
            
            return nolJam + std::to_string(this->jam) + ":" + nolMenit+ std::to_string(this->menit) + ":" +nolDetik+ std::to_string(this->detik);
        }
        int convertToSecond(){
            int hasil = this->detik + this->menit*60 + this->jam*3600;
            return hasil;
        }
        void secondToClock(int second){
            this->menit=second/60;
            this->detik=second%60;
            this->jam=this->menit/60;
            this->menit=this->menit%60;
        }
        Waktu cariDurasi(Waktu akhir){
             Waktu temp;

            int detikAwal = this->convertToSecond();
            int detikAkhir = akhir.convertToSecond();
            if(detikAkhir<detikAwal){
                detikAkhir+=86400;
            }
            int detikHasil = detikAkhir - detikAwal;

            

            temp.secondToClock(detikHasil);
            return temp;
        }
};

class Kendaraan{
    protected:
        std::string no;
        std::string jenis;
        Waktu datang;
        Waktu pulang;
    
    public:
        Kendaraan(){
            this->no = " ";
        }

        //Input
        void setNoKendaraan(std::string no){
            this->no = no;
        }

        void setJenis(std::string jenis){
            this->jenis = jenis;
        }

        void setWaktuDatang(Waktu datang){
            this->datang = datang;

        }

        void setWaktuPulang(Waktu pulang){
            this->pulang=pulang;
        }

        void inputKendaraan(){
            std::cout<<"\n--- INPUT KENDARAAN ---\n";
            std::cout<<"No Kendaraan : ";
            std::cin.ignore();
            std::getline(std::cin,this->no);

            std::cout<<"\n-- Jam Masuk Kendaraan --\n";
            this->datang.inputWaktu();

            std::cout<<"\n-- Jam Keluar Kendaraan --\n";
            this->pulang.inputWaktu();
        }

        //Output  
        std::string getNoKendaraan(){
            return this->no;
        }

        std::string getJenis(){
            return this->jenis;
        }

        Waktu getWaktudatang(){
            return this->datang;
        }

        Waktu getWaktuPulang(){
            return this->pulang;
        }

        //Proses 
        Waktu getLamaParkir(){
            return this->datang.cariDurasi(this->pulang);
        }

        int getLamaJam(){
        int hasil = 0;
            if(this->getLamaParkir().getMenit()>=10 || this->getLamaParkir().getJam()>=1){
                hasil = this->getLamaParkir().getJam();

                if( this->getLamaParkir().getMenit()>0 || this->getLamaParkir().getDetik()>0 ){
                    hasil +=1;
                }
           
            }
            return hasil;
        }

        virtual int getBiayaParkir(){return 0;}
   

};

class Mobil : public Kendaraan{

    public:
        Mobil():Kendaraan(){

            this->jenis = "Mobil";
        }

        int getBiayaParkir(){
            return this->getLamaJam()*3000;
        }
};

class Motor : public Kendaraan{
    public:
        Motor():Kendaraan(){
            this->jenis="Motor";
        }

         int getBiayaParkir(){
            return this->getLamaJam()*2000;
        }
};

class Truck : public Kendaraan{
    public:
        Truck():Kendaraan(){
            this->jenis="Truck";
        }

         int getBiayaParkir(){
            return this->getLamaJam()*10000;
        }
};

class LarikKendaraan{
    private:
        int ukuran;
        Kendaraan *p[100];
        
    
    public:
        LarikKendaraan(int ukuran){
            this->ukuran = ukuran;
        }
        void inputLarikParkir(){
            int num;
                for(int i=0;i<this->ukuran;i++){
                    system("cls");
                    std::cout<<"Kendaraan ke - " <<(i+1)<<"\n";
                    std::cout<<"Jenis Kendaraan : \n";
                    std::cout<<"1 . Mobil\n";
                    std::cout<<"2 . Motor\n";
                    std::cout<<"3 . Truck\n";
                    std::cout<<"Input....";
                    std::cin>>num;

                    switch (num)
                    {
                    case 1:
                        this->p[i] = new Mobil();
                        break;
                    
                     case 2:
                        this->p[i] = new Motor();
                        break;
                    
                     case 3:
                        this->p[i] = new Truck();
                        break;
                    
                    }

                    this->p[i]->inputKendaraan();    
                    system("cls");
            }
        }
        void cetakTabelParkir(){
            std::cout<<"\t\t\t\tRekapitulasi Biaya parkir PT Parkir Jaya\n";

            if(this->p[0]->getNoKendaraan()==" "){
                std::cout<<"Larik Kosong !\n";
            }

            else{
                int no = 1;
                std::cout<<"================================================================================================================================\n";
                std::cout<<"No\tNo Kendaraan\t\tJenis\tJam Masuk\tJam Pulang\tLama Parkir\tLama jam\tBiaya\n";
                std::cout<<"================================================================================================================================\n";

                for(int i=0;i<this->ukuran;i++){
                    if(this->p[i]->getNoKendaraan() == " "){
                        break;
                    }

                    else{
                        std::cout<<no<<"\t"<<
                        this->p[i]->getNoKendaraan() << "\t\t" << 
                        this->p[i]->getJenis()<< "\t" << 
                        this->p[i]->getWaktudatang().getWaktu() << "\t" << 
                        this->p[i]->getWaktuPulang().getWaktu() << "\t" <<   
                        this->p[i]->getLamaParkir().getWaktu() << "\t " << 
                        this->p[i]->getLamaJam() << "\t\t" << 
                        this->p[i]->getBiayaParkir() << "\t  \n";  
                    }
                    no++;
                }
                std::cout<<"================================================================================================================================\n";
            }
        }
        float totalBiaya(){
            float hasil = 0;
            for(int i=0;i<this->ukuran;i++){
                hasil+=this->p[i]->getBiayaParkir();
            }

            return hasil;
        }

};

int main()
{
    int ukuran;
    std::cout<<"Masukkan jumlah kendaraan = ";
    std::cin>>ukuran;

    LarikKendaraan lp(ukuran);

    system("cls");

    lp.inputLarikParkir();
    system("cls");
    lp.cetakTabelParkir();

    std::cout<<"Total Biaya adalah = "<<lp.totalBiaya();

    return 0;
}

File: SEMESTER3/PemrogramanBerorientasiObjek/Week10/TUGAS/Inheritance.py
# Nama          : Amir Salim , Andre Nathaniel Adipraja , Prames Ray lapian , Ibrahim Dafi Iskandar
# NPM           : 140810210015 , 140810200042 , 140810210059 , 140810210039
# Kelas         : A
# Tanggal       : 11 November 2022
# Nama Program  : Inheritance.py
# Deskripsi     : UTS SOAL 4 dengan turunan

from os import system

class Waktu:

    #Attribute
    __jam=0
    __menit=0
    __detik=0

    #Constructor
    def __init__(self, *args):
          if (len(args) == 3):
            self.__jam = int(args[0])
            self.__menit = int(args[1])
            self.__detik = int(args[2])
          
          elif(len(args)==0):
            self.__jam = int(0)
            self.__menit = int(0)
            self.__detik = int(0)

          else:
            print("False number of argument in constructor") 

    #Input Method
    def setJam(self,jam):
        self.__jam = int(jam)
    
    def setMenit(self,menit):
        self.__menit = int(menit)
    
    def setDetik(self,detik):
        self.__detik = int(detik)
    
    def inputWaktu(self):
       
        self.__jam = int(input("Masukkan jam : "))
        self.__menit = int(input("Masukkan menit : "))
        self.__detik = int(input("Masukkan detik : "))
    
    #Output Method
    def getJam(self):
        return self.__jam
    
    def getMenit(self):
        return self.__menit
    
    def getDetik(self):
        return self.__detik
    
    def getWaktu(self):
        nolJam =""
        nolMenit=""
        nolDetik=""

        if(self.__jam<10):
            nolJam="0"
            
        if(self.__menit<10):
            nolMenit="0"
            
        if(self.__detik<10):
            nolDetik="0"

        return nolJam + str(self.__jam) + ":" + str(nolMenit)+ str(self.__menit) + ":" +nolDetik+ str(self.__detik)
    
    #Proses

    def convertToSecond(self):
        hasil = self.__detik + (int(60) * self.__menit) + (int(3600) * self.__jam)
        return hasil 
    
    def secondToClock(self,second:int):
        self.__menit = int(second/60)
        self.__detik = int(second%60)
        self.__jam = int(self.__menit/60)
        self.__menit = int(self.__menit%60)
    
    def cariDurasi(self,akhir):
        temp = Waktu()

        detikAwal = self.convertToSecond()
        detikAkhir = akhir.convertToSecond()

        if(detikAkhir<detikAwal):
            detikAkhir+=86400
        
        detikHasil = detikAkhir - detikAwal
        temp.secondToClock(detikHasil)

        return temp

#-----------------------------------------------------------------------------------------------------------------------------------------

class Kendaraan:
    _no = " "
    _jenis= " "
    _datang = Waktu()
    _pulang = Waktu()

    #Constructor
    def __init__(self):
        self._no = " "
        self._jenis = " "
        self._datang = Waktu(0,0,0)
        self._pulang = Waktu(0,0,0)


    #Input
    def setNoKendaraan(self,no):
        self._no = no
    
    def setJenis(self,jenis):
        self._jenis=jenis
    
    def setWaktuDatang(self,datang):
        self._datang = datang
    
    def setWaktuPulang(self,pulang):
        self._pulang = pulang
    
    def inputKendaraan(self):
        print("\n--- INPUT KENDARAAN---")
        self._no = input("No Kendaraan : ")

        print("\n--- Jam Masuk Kendaraan ---")
        self._datang.inputWaktu()

        print("\n--- Jam Keluar Kendaraan ---")
        self._pulang.inputWaktu()
    
    #Output
    def getNoKendaraan(self):
        return self._no

    def getJenis(self):
        return self._jenis
    
    def getWaktuDatang(self):
        return self._datang
    
    def getWaktuPulang(self):
        return self._pulang   
    def getLamaParkir(self):
        return self._datang.cariDurasi(self._pulang)
    
    def getLamaJam(self):
        hasil = int(0)
        if(self.getLamaParkir().getMenit()>=10 or self.getLamaParkir().getJam()>=1):
            hasil = self.getLamaParkir().getJam()

            if(self.getLamaParkir().getMenit()>0 or self.getLamaParkir().getDetik()>0):
                hasil = hasil + 1
        
        return hasil
    
    
    def getBiayaParkir(self):
        ...
#-----------------------------------------------------------------------------------------------------------------------------------------   

class Motor(Kendaraan):
    def __init__(self):
        super().__init__()
        self._jenis = "Motor"
    
    def getBiayaParkir(self):
        return self.getLamaJam() * 2000

#-----------------------------------------------------------------------------------------------------------------------------------------
class Mobil(Kendaraan):
    def __init__(self):
        super().__init__()
        self._jenis = "Mobil"
    
    def getBiayaParkir(self):
        return self.getLamaJam() * 3000
#-----------------------------------------------------------------------------------------------------------------------------------------
class Truck(Kendaraan):
    def __init__(self):
        super().__init__()
        self._jenis = "Truck"
    
    def getBiayaParkir(self):
        return self.getLamaJam() * 10000
#-----------------------------------------------------------------------------------------------------------------------------------------
class LarikKendaraan:
    __ukuran=int(0)
    __p = []

    #Constructor
    def __init__(self,ukuran):
        self.__ukuran = ukuran
    
    #Input
    def inputLarikParkir(self):
        i = int(0)
        num = int(0)
        while(i<self.__ukuran):
            system("cls")
            print("Kendaraan ke - ",(i+1))
            
            print("Jenis Kendaraan : ")
            print("1 . Mobil")
            print("2 . Motor")
            print("3 . Truck")
            num = int(input("Input...."))
            

            match num:
                case 1:
                    obj = Motor()
                
                case 2:
                    obj = Mobil()
                
                case 3:
                    obj = Truck()
                

            
            obj.inputKendaraan()
            self.__p.append(obj)
            system("cls")
            i=i+1
    
    #Output
    def cetakTabelParkir(self):
        print("\t\t\t\tRekapitulasi Biaya parkir PT Parkir Jaya\n")
        if(self.__p[0].getNoKendaraan()==" "):
            print("Larik Kosong ! ")
        
        else:
            no = int(1)
            i = int(0)
            print("================================================================================================================================")
            print("No\tNo Kendaraan\t\tJenis\tJam Masuk\tJam Pulang\tLama Parkir\tLama jam\tBiaya");
            print("================================================================================================================================")
            while(i<self.__ukuran):
                if(self.__p[i].getNoKendaraan() == " "):
                    break

                else:
                    print(
                    no,"\t",
                    self.__p[i].getNoKendaraan(),"\t\t", 
                    self.__p[i].getJenis(),"\t", 
                    self.__p[i].getWaktuDatang().getWaktu(),"\t", 
                    self.__p[i].getWaktuPulang().getWaktu(),"\t",   
                    self.__p[i].getLamaParkir().getWaktu(),"\t ", 
                    self.__p[i].getLamaJam(),"\t\t", 
                    self.__p[i].getBiayaParkir(),"\t  "  
                         )
                no = no+1;
                i = i+1
            print("================================================================================================================================")
    
    #Proses
    def totalBiaya(self):
        hasil = float(0)
        i = int(0)
        while(i<self.__ukuran):
            hasil = hasil + self.__p[i].getBiayaParkir()
            i = i + 1
        return hasil


#-----------------------------------------------------------------------------------------------------------------------------------------

uk = int(input("Masukkan Jumlah Kendaraan : "))

system("cls")

lp = LarikKendaraan(uk)

lp.inputLarikParkir()
system("cls")
lp.cetakTabelParkir()

print("Total Biaya Parkir adalah = ",lp.totalBiaya())